src/ngraph/runtime/gpu/cublas_emitter.cpp

```diff
@@ -0,0 +1,240 @@
+//*****************************************************************************
+// Copyright 2017-2018 Intel Corporation
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//*****************************************************************************
+#include <algorithm>
+#include <iostream>
+#include <memory>
+#include <sstream>
+#include <stdexcept>
+
+#include "ngraph/log.hpp"
+#include "ngraph/runtime/gpu/cublas_emitter.hpp"
+#include "ngraph/runtime/gpu/gpu_emitter.hpp"
+#include "ngraph/runtime/gpu/gpu_invoke.hpp"
+#include "ngraph/runtime/gpu/gpu_primitive_emitter.hpp"
+#include "ngraph/runtime/gpu/gpu_runtime_context.hpp"
+#include "ngraph/runtime/gpu/gpu_util.hpp"
+#include "ngraph/runtime/gpu/type_info.hpp"
+#include "ngraph/util.hpp"
+
+using namespace ngraph;
+using namespace std;
+
+runtime::gpu::CUBLASEmitter::CUBLASEmitter(GPUPrimitiveEmitter* emitter, GPURuntimeContext* ctx)
+    : m_primitive_emitter(emitter)
+{
+    m_ctx = ctx;
+}
+
+size_t runtime::gpu::CUBLASEmitter::build_dot(const element::Type& dtype,
+                                              const Shape& arg0_shape,
+                                              const Shape& arg1_shape,
+                                              const Shape& out_shape,
+                                              size_t reduction_axes)
+{
+    std::stringstream ss;
+    ss << "dot_op"
+       << "_dtype_" << dtype.c_type_string() << "_reduction_axes_count_" << reduction_axes;
+    std::string hash = ss.str() + "_i_" + join(arg0_shape, "_") + "_i_" + join(arg1_shape, "_") +
```

Comment:
Here you are hashing the full shapes of the input tensors and the output tensor. Does the hash need to be so restrictive? Maybe it could depend only on the rank? Would be nice if so, so that we don't have to have a unique primitive for each tensor shape. Maybe it can't, but if you could review the kernel and see if there isn't a way to simplify it the hash can be rank dependent only, that would be great.

Reply:
at least one of it is redundant between input0, input1, output and reduction axes.

Reply:
For now, just remove the output shape in the hash. We discussed possibly breaking up the primitive build function into separate calls for blas-1/2/3 operations so that they can be cached based on rank rather than tensor shape.